File: seoyeong/solved/BOJ_9012.py
t = int(input())

for tc in range(1, t+1):
    text = input()

    stack = [] # 스택 생성
    result = 'YES' # 디폴트 = YES

    for i in text:
        if i == '(': 
            stack.append(i) # '('를 만나면 스택에 어펜드
        else:
            if len(stack) == 0:
                result = 'NO' #')'를 만났는데 스택에 남아있는게 없으면 결과 = 'NO'
                break # 더 볼 필요가 없으니 break
            else:
                stack.pop() # 만약 스택에 남아있는게 있으면 팝
    if len(stack) != 0: 
        result = 'NO' # 다 돌았는데 스택에 남아있는게 있으면 짝이 맞지 않다는 것. 결과 = 'NO'

    print(result)

File: kyungrin/solved/BOJ_9012.java
package kyungrin.solved;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

/**
 * 메모리: 14316 kb
 * 시간: 108 ms
 *
 * 2 <= 각 입력의 길이 <= 50
 * <p>
 * 여는 괄호 STK 보관. 닫는 괄호는 POP 1. 닫는 괄호를 받았는데 STK이 EMPTY할 시 UVPS 2. 모든 순회가 끝났는데 남아있으면 UVPS
 * <p>
 * 그외의 경우는 VPS이다.
 **/

// 디버깅 시 AI 사용
public class BOJ_9012 {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(br.readLine());

    StringBuilder sb = new StringBuilder();
    for (int i = 0; i < N; i++) { // N회, 0 ~ N-1회 반복
      Stack<Character> stk = new Stack<>();
      String temp = br.readLine();
      boolean isUVPS = false;

      for (int j = 0; j < temp.length(); j++) {

        char p = temp.charAt(j); // AI: j 대신 i를 넣고 있었다.
        if (p == '(') {
          stk.push(p);
        } else if (p == ')') {
          if (stk.isEmpty()) {
            isUVPS = true;
            break;
          }
          stk.pop();
        }
      }

      if (!stk.isEmpty() || isUVPS) { // or 연산자(||)가 아닌 비트마스크 연산자(|)를 쓰고 있었다.
        sb.append("NO").append("\n");
      } else {
        sb.append("YES").append("\n");
      }
    }

    System.out.println(sb);
  }
}
